File: lib/selfAdaptiveMA.js
function sum(data)
{
	var sum = 0;
	for (var i = 0; i < data.length; ++i)
	{
		sum += data[i];
	}
	return sum;
}

function ma(data, n)
{
	var dataNum = data.length;
	var output = Array();
	var i;
	for (i = 0; i <= n; ++i){
		output[i] = sum(data.slice(0, i)) / ( i + 1);
	}
	for (i = n+1; i < dataNum; ++i){
		dir = Math.abs(data[i-1]-data[i-n-1]);
		vir = 0;
		for (var k = i - n; k < i; ++k){
			vir = vir + Math.abs(data[k - 1] - data[k - 2]);
		}
		var er = dir / vir;
		var sc = er*(2/3-2/29)+2/29;
		var c = sc*sc;
		output[i-1] = output[i-2]+c*(data[i-1]-output[i-2]);

	}
	return output
}

var ttt = Array(100, 200, 300, 400, 555, 298, 2938, 2983, 2398, 2983, 2389823, 298);
console.log(ma(ttt, 2));

File: lib/selfAdaptiveMA.lua
require('math')
require('lib/common')

function selfAdaptiveMA( data, N, fastLen, showLen )
	--%SELFADAPTIVEMA 自适应均线
	--%   
	fastLen = fastLen or 2
	slowLen = slowLen or 10
	local fast = 2 / (fastLen + 1)
	local slow = 2 / (slowLen + 1)
	local numRow = #data;
	local output = {}
	output [numRow] = 0

	--%output(1) = data(1);

	for i = 1,N do
		output[i] = Matlab.sum(Matlab.arrayslice(data, 1, i)) / i;
	end

	for i = N+1,numRow do
		local dir = math.abs(data[i]-data[i-N]);
		local volatility = 0;
		for k = i-N+1,i do
			volatility = volatility + math.abs(data[k]-data[k-1]);
		end
		local er = dir/volatility;
		local sc = er*(fast - slow) + slow;
		local c = sc*sc;
		output[i] = output[i-1] + c * (data[i]-output[i-1]);
		
	end
	return output
end
